feat(rust): add reset command implementation

michealkeines opened the pull request:
# Added

new reset command that does the same `node delete --force --all`

# Output

![reset-impl](https://user-images.githubusercontent.com/38526063/189470806-161e04a4-e619-44f7-9997-c4b8289df666.gif)

## Checks

<!-- To help us review and merge this pull request quickly, please confirm the following:  -->

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

<!-- Looking forward to merging your contribution!! -->

mrinalwadhwa reviewed (changes requested):
@michealkeines thank you for spending time on another great contribution.

I think it would better to keep most of this code in one place ... maybe a `node::delete_all_with_force` that can be called from both locations.

michealkeines commented:
I have moved the following function into utils

`delete_node`
`delete_node_pid`
`delete_node_config`

added a new function `delete_nodes_with_options`, this can be called in both `node delete` and `reset`

# outputs:

## delete node with name

![image](https://user-images.githubusercontent.com/38526063/189476364-ea80a4c6-704c-422e-aeae-f6acdab0e616.png)

## delete node with force

![image](https://user-images.githubusercontent.com/38526063/189476392-1ce48ed8-5979-43c4-bf5a-745a4ffcfc3c.png)

## delete node with force and all

![image](https://user-images.githubusercontent.com/38526063/189476420-8b50910f-15bc-40e0-829b-ae341f5bcf9d.png)

## reset

![image](https://user-images.githubusercontent.com/38526063/189476455-0b03e82f-8ce9-4b17-b3ab-8bd996a06f9f.png)

## reset with yes

![image](https://user-images.githubusercontent.com/38526063/189476482-2584c395-b565-4630-b7cd-3451690ef5dd.png)

michealkeines commented:
Also, in latest dev branch

delete command prints `Deleted node` even for Node that doesn't exist

![image](https://user-images.githubusercontent.com/38526063/189476724-9c1a07f6-238d-4e65-b538-cd2108a30a67.png)

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/delete.rs:
```diff
@@ -31,94 +29,38 @@ impl DeleteCommand {
+       If --Force was set and not --All, we fall back to this else condition with `delete_node`
+       If None, we fall back to this else condition with `delete_node`
+    */
+    if cmd.all && res.is_ok() {
```
You should match the `res` inside this `if` to bubble up the error. Otherwise you would be ignoring the errors that might happen inside `delete_nodes_with_options` that were previously checked before this PR.

This could be rewritten as (note that I didn't pass the `cmd.all` argument to the function):
```rust
if cmd.all {
    delete_nodes_with_options(opts.clone(), cmd.force)?;
}
```

Also, I think the comments above the `if` are not necessary, since the code is pretty simple and self-explanatory at this point.

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/util.rs:
```diff
@@ -144,3 +148,87 @@ pub async fn delete_embedded_node(cfg: &OckamConfig, name: &str) {
 }
+
+pub fn delete_nodes_with_options(opts: CommandGlobalOpts, all: bool, force: bool) -> Result<()> {
+    if all {
```
You could remove the `all` argument, as it has no effect. After that change, renaming the function to `delete_all_nodes` might be more explicit as to what it does.

michealkeines replied:
updates as per your comments

adrianbenavides commented on implementations/rust/ockam/ockam_command/src/node/delete.rs:
```diff
@@ -31,94 +29,38 @@ impl DeleteCommand {
+       If None, we fall back to this else condition with `delete_node`
+    */
+    if cmd.all && res.is_ok() {
+        // If force is not enabled
+        if !cmd.force {
+            // persist updates to the config file
+            if let Err(e) = opts.config.persist_config_updates() {
+                eprintln!("Failed to update config file. You might need to run the command with --force to delete all config directories");
+                return Err(crate::Error::new(crate::exitcode::IOERR, e));
```
Why not keep this inside the `delete_nodes_with_options` function?

michealkeines replied:
moved it to `delete_all_nodes()`

michealkeines commented:
I removed this `Error` import in lib.rs, as per warning from cargo

![image](https://user-images.githubusercontent.com/38526063/189724233-9884aa54-bbdb-407b-ae13-72bf0f1e956e.png)

michealkeines commented:
Hi @adrianbenavides I have updated the code as per your comments, please review it.

adrianbenavides reviewed (dismissed):
Great! Thanks @michealkeines for looking into this and going through all the comments! :rocket:

adrianbenavides commented:
@michealkeines can you fix that tiny conflict, please? We can merge this PR after that :+1:

michealkeines committed:
feat(rust): add `reset` command, refactor `node delete`

michealkeines commented:
I have rebased it to latest develop @adrianbenavides

michealkeines commented:
Hi, @adrianbenavides could you please check this again?

adrianbenavides commented:
@mergifyio rebase

adrianbenavides reviewed (approved):
Looks ready, thanks again @michealkeines!

mrinalwadhwa commented:
@michealkeines Thank you for another awesome contribution 🥳 🎈
